cirq/protocols/measurement_key.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Protocol for object that have measurement keys."""
+
+from typing import Any
+
+from typing_extensions import Protocol
+
+# This is a special indicator value used by the inverse method to determine
+# whether or not the caller provided a 'default' argument.
+RaiseTypeErrorIfNotProvided = ([],)  # type: Any
+
+
+class SupportsMeasurementKey(Protocol):
+    """An object that is a measurement and has a measurement key.
+
+    Measurement keys are used in referencing the results of a measurement.
+
+    Note: Measurements, in contrast to general quantum channels, are
+    distinguished by the recording of the quantum operation that occurred.
+    That is a general quantum channel may enact the evolution
+        \rho -> \sum_k A_k \rho A_k^\dagger
+    where as a measurement enacts the evolution
+        \rho -> A_k \rho A_k^\dagger
+    conditional on the measurement outcome being k.
+    """
+
+    def _measurement_key_(self) -> str:
+        """Return the key that will be used to identify this measurement.
+
+        When a measurement occurs, either on hardware, or in a simulation,
+        this is the key value under which the results of the measurement
+        will be stored.
+        """
+
+
+def measurement_key(
```

Comment:
What do you think about slightly generalizing this so that it works as a way of getting the key for measurement operations, but also as a way to get the canonical key for e.g. a list of qubits?

Reply:
I think lets start with "measurement key" but it could be that we want the general method at some point.